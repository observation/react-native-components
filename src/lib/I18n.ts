import { I18nManager } from 'react-native'

import Countries from 'i18n-iso-countries'
import { I18n as I18nJs } from 'i18n-js'
import 'moment/min/locales'
import { Scope, TranslateOptions } from 'i18n-js/typings/typing'
import moment from 'moment'
import * as RNLocalize from 'react-native-localize'

import Log from './Log'

Countries.registerLocale(require('i18n-iso-countries/langs/en.json'))
Countries.registerLocale(require('i18n-iso-countries/langs/nl.json'))
Countries.registerLocale(require('i18n-iso-countries/langs/de.json'))
Countries.registerLocale(require('i18n-iso-countries/langs/es.json'))
Countries.registerLocale(require('i18n-iso-countries/langs/fr.json'))

interface SupportedLanguages {
  en: Object
  nl: Object
  de: Object
  fr: Object
  es: Object
}

interface Api {
  setLanguage(language: string): void
}

/** Wrapper class around i18n-js, adding a few more methods */
class I18n {
  i18nJs = new I18nJs()

  constructor(private translations: SupportedLanguages, private api: Api) {
    this.i18nJs.translations = translations
    this.i18nJs.enableFallback = false
    this.configure()
  }

  setLocale = (locale: string) => {
    Log.debug('i18n:setLocale', locale)
    this.i18nJs.locale = locale
    moment.locale(locale)
    this.api.setLanguage(locale)
  }

  get locale() {
    return this.i18nJs.locale
  }

  /** Handle setting an initial or changed locale */
  configure = () => {
    const fallback = { languageTag: 'en', isRTL: false }
    const languages = Object.keys(this.translations)
    const { languageTag, isRTL } = RNLocalize.findBestAvailableLanguage(languages) || fallback
    Log.debug('i18n:configure', languageTag)
    I18nManager.forceRTL(isRTL) // optional, you might not want to handle RTL
    this.setLocale(languageTag)
  }

  getCountryFromLocale = () => {
    Log.debug('i18n:getCountryFromLocale')

    const fallback = 'NL'
    const locales = RNLocalize.getLocales()
    return locales.length > 0 ? locales[0].countryCode : fallback
  }

  getLanguageFromLocale = () => {
    Log.debug('i18n:getLanguageFromLocale')

    const locales = RNLocalize.getLocales()
    return locales.length > 0 ? locales[0].languageCode : null
  }

  country = (code: string) => Countries.getName(code, this.locale, { select: 'official' })

  t = (key: Scope, options?: TranslateOptions | undefined) => this.i18nJs.t(key, options)
}

export default I18n
export type { Api }
