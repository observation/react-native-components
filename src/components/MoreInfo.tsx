import React from 'react'
import { StyleProp, ViewStyle } from 'react-native'

import { IconInfo } from './Icon'
import IconText from './IconText'
import I18n from '../lib/I18n'
import textStyle from '../styles/text'
import theme from '../styles/theme'

type Props = {
  onPress?: () => void
  containerStyle?: StyleProp<ViewStyle>
  testID?: string
  i18n: I18n
}

const MoreInfo = ({ onPress, containerStyle, i18n }: Props) => (
  <IconText
    icon={<IconInfo size={theme.icon.size.medium} />}
    text={i18n.t('About the species')}
    style={{
      containerStyle,
      textStyle: textStyle.link,
    }}
    onPress={onPress}
  />
)

export default MoreInfo
