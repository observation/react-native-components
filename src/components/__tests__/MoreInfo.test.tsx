import React from 'react'

import { render, fireEvent } from '@testing-library/react-native'

import I18n from '../../lib/I18n'
import MoreInfo from '../MoreInfo'

const onPress = jest.fn()
const i18n = new I18n(
  { en: { 'About the species': 'About the species' }, nl: {}, fr: {}, de: {}, es: {} },
  { setLanguage: () => {} },
)
describe('MoreInfo', () => {
  test('Rendering', () => {
    const { toJSON } = render(<MoreInfo onPress={onPress} containerStyle={{ flex: 1 }} i18n={i18n} />)
    expect(toJSON()).toMatchSnapshot()
  })

  test('Click', async () => {
    const { getByText } = render(<MoreInfo onPress={onPress} containerStyle={{ flex: 1 }} i18n={i18n} />)
    await fireEvent.press(getByText('About the species'))
    expect(onPress).toBeCalled()
  })
})
