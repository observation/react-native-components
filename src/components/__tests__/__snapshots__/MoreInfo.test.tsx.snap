// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoreInfo Rendering 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
>
  <View
    style={
      [
        {
          "flexDirection": "row",
        },
        {
          "flex": 1,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "justifyContent": "center",
            "marginRight": 8,
          },
          undefined,
        ]
      }
    >
      <Icon
        color="#0066B1"
        name="circle-info"
        size={14}
        type="light"
      />
    </View>
    <Text
      style={
        [
          {
            "flexShrink": 1,
          },
          {
            "color": "#0066B1",
            "fontFamily": "Ubuntu",
            "fontSize": 14,
            "fontStyle": "normal",
            "fontWeight": "normal",
            "lineHeight": 20,
          },
        ]
      }
    >
      About the species
    </Text>
  </View>
</View>
`;
